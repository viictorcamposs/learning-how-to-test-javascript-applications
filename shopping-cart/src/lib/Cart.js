import { find, remove } from 'lodash'

export class Cart {
  items = []

  add(item) {
    const itemToFind = { product: item.product }

    if (find(this.items, itemToFind)) {
      remove(this.items, itemToFind)
    }

    this.items.push(item)
  }

  remove(product) {
    remove(this.items, { product })
  }

  getTotal() {
    return !this.items.length < 1
      ? this.items.reduce(
          (accumulator, { quantity, product: { price } }) =>
            accumulator + quantity * price,
          0,
        )
      : 0
  }

  summary() {
    return {
      items: this.items,
      total: this.getTotal(),
    }
  }

  checkout() {
    const { items, total } = this.summary()

    this.items = []

    return {
      items,
      total,
    }
  }
}
