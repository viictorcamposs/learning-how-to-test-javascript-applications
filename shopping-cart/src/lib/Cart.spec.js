import { Cart } from './Cart'

describe('Cart', () => {
  let cart

  const product = {
    title: 'Adidas running shoes - men',
    price: 35388,
  }

  const product2 = {
    title: 'Adidas running shoes - women',
    price: 41872,
  }

  beforeEach(() => {
    cart = new Cart()
  })

  describe('getTotal()', () => {
    it('should return 0 when function getTotal is executed in a newly created instance', () => {
      expect(cart.getTotal().getAmount()).toEqual(0)
    })

    it('should multiply quantity and price and receive the total amount', () => {
      const item = {
        product,
        quantity: 2,
      }

      cart.add(item)

      expect(cart.getTotal().getAmount()).toEqual(70776)
    })

    it('should ensure no more than one product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product,
        quantity: 1,
      })

      expect(cart.getTotal().getAmount()).toEqual(35388)
    })

    it('should update total when a product gets included and then removed', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product: product2,
        quantity: 1,
      })

      cart.remove(product)

      expect(cart.getTotal().getAmount()).toEqual(41872)
    })
  })

  describe('checkout()', () => {
    it('should return an object with the total and the list of items when summary() is called', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product: product2,
        quantity: 1,
      })

      expect(cart.summary()).toMatchSnapshot()
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0)
    })

    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product: product2,
        quantity: 1,
      })

      expect(cart.checkout()).toMatchInlineSnapshot(`
        Object {
          "items": Array [
            Object {
              "product": Object {
                "price": 35388,
                "title": "Adidas running shoes - men",
              },
              "quantity": 2,
            },
            Object {
              "product": Object {
                "price": 41872,
                "title": "Adidas running shoes - women",
              },
              "quantity": 1,
            },
          ],
          "total": 112648,
        }
      `)
    })

    it('should reset the cart when is called', () => {
      cart.add({
        product: product2,
        quantity: 1,
      })

      cart.checkout()

      expect(cart.items).toEqual([])
    })
  })
})
