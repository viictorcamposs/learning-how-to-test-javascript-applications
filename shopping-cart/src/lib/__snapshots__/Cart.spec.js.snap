// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items when summary() is called 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas running shoes - women",
      },
      "quantity": 1,
    },
  ],
  "total": 112648,
}
`;
